src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java

```diff
@@ -0,0 +1,728 @@
+
+import javax.annotation.Nonnull;
+
+import java.util.*;
+
+public class PatternAcceptor {
+
+    public final String pattern;
+    public final boolean unicode;
+
+    private static final String[] decimalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"};
+    private static final String[] octalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7"};
+    private static final String[] hexDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f", "A", "B", "C", "D", "E", "F"};
+    private static final String syntaxCharacters = "^$\\.*+?()[]{}|";
+    private static final String[] syntaxCharacterArray = "^$\\.*+?()[]{}|".split("");
+    private static final String extendedSyntaxCharacters = "^$.*+?()[|";
+    private static final String[] controlCharacters = new String[]{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};
+
+    private static HashMap<String, Integer> constructControlEscapeCharacterValues() {
+        HashMap<String, Integer> map = new HashMap<>();
+        map.put("f", (int) '\f');
+        map.put("n", (int) '\n');
+        map.put("r", (int) '\r');
+        map.put("t", (int) '\t');
+        map.put("v", 0x11); // \v in javascript
+        return map;
+    }
+
+    private static final HashMap<String, Integer> controlEscapeCharacterValues = constructControlEscapeCharacterValues();
+
+    private static final String[] controlEscapeCharacters = controlEscapeCharacterValues.keySet().toArray(new String[0]);
+    
+    private class Context {
+        private int index;
+        private ImmutableSet<String> backreferenceNames;
+        private ImmutableSet<String> groupingNames;
+        private ImmutableSet<Integer> backreferences;
+        private int capturingGroups;
+
+        private Context(@Nonnull Context context) {
+            this.index = context.index;
+            this.backreferenceNames = context.backreferenceNames;
+            this.groupingNames = context.groupingNames;
+            this.backreferences = context.backreferences;
+            this.capturingGroups = context.capturingGroups;
+        }
+
+        public Context() {
+            this.index = 0;
+            this.backreferenceNames = ImmutableSet.emptyUsingEquality();
+            this.groupingNames = ImmutableSet.emptyUsingEquality();
+            this.backreferences = ImmutableSet.emptyUsingEquality();
+            this.capturingGroups = 0;
+        }
+
+        public boolean addGrouping() {
+            this.capturingGroups++;
+            return true;
+        }
+
+        public void backreference(int num) {
+            this.backreferences = this.backreferences.put(num);
+        }
+
+        public boolean verifyBackreferences() {
+            if (unicode) {
```

Comment:
This is an instance var, so let's use `this.` so we know it's not static or local.